pkg/api/helpers/download.go

```diff
@@ -0,0 +1,60 @@
+package helpers
+
+import (
+	"github.com/treeverse/lakefs/pkg/api"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+)
+
+// ClientDownload downloads a file using client-side ("direct") access to underlying storage.
+// It requires credentials both to lakeFS and to underlying storage, but considerably reduces
+// the load on the lakeFS server.
+func ClientDownload(ctx context.Context, client api.ClientWithResponsesInterface, repoID, ref, filePath string) (*api.ObjectStats, io.ReadCloser, error) {
+	resp, err := client.StatObjectWithResponse(ctx, repoID, ref, &api.StatObjectParams{
+		Path: filePath,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("stat object: %w", err)
+	}
+	if err := ResponseAsError(resp); err != nil {
+		return nil, nil, fmt.Errorf("get object URI: %w", err)
+	}
+
+	physicalAddress:= resp.JSON200.PhysicalAddress
+	parsedAddress, err := url.Parse(physicalAddress)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse physical address URL %s: %w", physicalAddress, err)
+	}
+
+	// TODO(ariels): plug-in support for other protocols
+	if parsedAddress.Scheme != "s3" {
```

Comment:
I've seen this elsewhere too, I think the supported scheme knowledge shouldn't spread in more than 1 location.

Reply:
My plan:
1. Pull this.
2. Write additional support for other backing stores.
3. As part of :arrow_up: :arrow_up: :arrow_up: :arrow_up:  write some factory with `get{Up,Down}Loader` methods.
4. Automate on _that_, plus an explicit additional "auth failed" error (to indicate an indirect retry may work).

Again, this is too early to write the factory I think.